### ISSUE TITLE
linker: undefined reference to `test<bool>::~test() // when method of template class is not inline, and that class is used with same template args in more than one .cpp

### ISSUE THREAD

**Issue by deemetrius** [brechtsanders/winlibs_mingw#83]
There is minimal example of two files, using modules:

lib.cpp
```
export module test;

export template <typename T>
struct test {
	T v_;

	test(T v) : v_(v) {}
	
	//inline
	~test() {}
};

export struct omg {
	using tt = test<bool>;

	static tt go(tt t) {
		return t;
	}
};
```

main.cpp
```
import test;

int main() {
	test<bool> t = false;
	return 0;
}
```

Here compile messages:
```
Perform "_build" task:
x86_64-w64-mingw32-g++.exe -Wall -O3 -std=c++20 -m64 -c D:\prog\Dropbox\cpp\vs-code/src/lib.cpp -o D:\prog\Dropbox\cpp\vs-code/obj/src/lib.o -fmodules-ts
x86_64-w64-mingw32-g++.exe -Wall -O3 -std=c++20 -m64 -c D:\prog\Dropbox\cpp\vs-code/src/main.cpp -o D:\prog\Dropbox\cpp\vs-code/obj/src/main.o -fmodules-ts
Perform "_link" task:
x86_64-w64-mingw32-g++.exe -o D:\prog\Dropbox\cpp\vs-code/bin/test.exe D:\prog\Dropbox\cpp\vs-code/obj/src/lib.o D:\prog\Dropbox\cpp\vs-code/obj/src/main.o -s -m64 -fmodules-ts
d:/prog/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/11.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:\prog\Dropbox\cpp\vs-code/obj/src/main.o:main.cpp:(.text.startup+0x14): undefined reference to `test<bool>::~test()'
collect2.exe: error: ld returned 1 exit status
```

But when I remove static method `omg::go()` then all fine.

Or if I change `export struct omg { using tt = test<bool>;` to `export struct omg { using tt = test<int>;` then again all fine.

Other option is to make `inline test::~test()` so have to make all methods of my template class inline in case those are called from more than one translation unit.

**Comment by deemetrius** [comment]
versions:
```
UCRT runtime
GCC 11.2.0 + LLVM/Clang/LLD/LLDB 13.0.0 + MinGW-w64 9.0.0 (UCRT) - release 2
```

used in os windows 10 (russian)

**Comment by brechtsanders** [comment]
The whole modules thing is quite new, and I haven't studied yet how to use it.

What are you trying to achieve defining a `static` member of a struct marked `export`?
I don't see how that makes sense.
As I understand it `static` means you want to restrict access to only local to this `.cpp` file while `export` means you want to share with others.

Have you tried reproducing this on other operating systems with GCC 11.2.0 (e.g. Linux)?

**Comment by brechtsanders** [comment]
I tried this on Linux (Debian bookworm) with GCC 11.2.0 and also got:
```
/usr/bin/ld: main.o: in function `main':
main.cpp:(.text.startup+0xf): undefined reference to `test<bool>::~test()'
collect2: error: ld returned 1 exit status
```

So this is not related in any way to MinGW-w64.

Instead, you should ask the GCC developers. You can do that by creating a bug report at https://gcc.gnu.org/bugs/

**Comment by deemetrius** [comment]
> What are you trying to achieve defining a `static` member of a struct marked `export`?
> I don't see how that makes sense.

The thing is to instantiate exported class `test<bool>` with same template arg `bool` in several translation units.
Here one is in `omg::go()` from 'lib.cpp' and the second is in `main()` from 'main.cpp' ;

I used static method of class because it can be inlined by compiler in some cases.
But it can be rewriten to just global function.

> I tried this on Linux (Debian bookworm) with GCC 11.2.0

Thanks.

> So this is not related in any way to MinGW-w64.
> Instead, you should ask the GCC developers. You can do that by creating a bug report at https://gcc.gnu.org/bugs/

Yes, I posted bug report to gcc:
[https://gcc.gnu.org/bugzilla/show_bug.cgi?id=104040](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=104040)

Hope they will fix it. This is important to me. I not want to get back to `#include` in my project. I like `import` now.
